consumer/mad/mad-0.14.2b/madplay.1
.\"
.\" mad - MPEG audio decoder
.\" Copyright (C) 2000-2001 Robert Leslie
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.\" $Id: madplay.1,v 1.36 2001/11/09 00:58:44 rob Exp $
.\"
.TH MADPLAY 1 "02 November 2001" "MAD" "MPEG Audio Decoder"
.SH NAME
madplay \- decode and play MPEG audio stream(s)
.SH SYNOPSIS
.B madplay
.RI [ options ]
.I file
\&...
.br
.B madplay
.RI [ options ]
\fB\-o\fR [\fItype\fB:\fR]\fIpath\fR
.I file
\&...
.SH DESCRIPTION
.B madplay
is a command-line MPEG audio decoder and player based on the MAD library
.RI ( libmad ).
.PP
MAD is a high-quality MPEG audio decoder. It currently supports MPEG-1 and the
MPEG-2 extension to Lower Sampling Frequencies, as well as the so-called
MPEG\ 2.5 format. All three audio layers (Layer\ I, Layer\ II, and Layer\ III
a.k.a. MP3) are fully implemented.
.PP
Among the special features of MAD are 24-bit PCM resolution and 100%
fixed-point (integer) computation. Since MAD is implemented entirely without
the use of floating point arithmetic, it performs especially well on
architectures without an FPU.
.PP
MAD does not yet support MPEG-2 multichannel audio (although it should be
backward compatible with such streams) nor does it currently support AAC.
.PP
By default
.B
madplay
reads and decodes one or more input
.IR file s
containing MPEG audio data and plays them on the native audio device. If the
input file is a single dash (\-), data is read from standard input.
.PP
Decoded output may optionally be redirected to a file instead of being played
on the audio device by using the
.B \-o
.RB ( \-\-output )
option.
.PP
For each
.IR file ,
.B madplay
will also attempt to read and display ID3 tag information. If data is being
read from standard input, ID3 tags will be displayed as they are encountered.
The supported tag versions are ID3v1, ID3v1.1, ID3v2.2, ID3v2.3, and ID3v2.4.
.SH OPTIONS
.SS Verbosity
.TP
.BR \-v " or " \-\-verbose
Show information about the stream during decoding, including playing time,
audio layer, bit rate, sampling frequency, and stereo mode.
.TP
.BR \-q " or " \-\-quiet
Do not show any information during decoding, except warnings.
.TP
.BR \-Q " or " \-\-very\-quiet
Do not show any information during decoding, including warnings.
.SS Decoding
.TP
.B \-\-downsample
Reduce the output sampling frequency 2:1. This also reduces the computational
overhead of the decoder.
.TP
.BR \-i " or " \-\-ignore\-crc
Ignore CRC information in the audio stream. This causes frames with CRC errors
to be decoded and played anyway. This option is not recommended, but since
some encoders have been known to generate bad CRC information, this option is
a work-around to play streams from such encoders.
.SS Audio Output
.TP
\fB\-o\fR or \fB\-\-output=\fR[\fItype\fB:\fR]\fIpath\fR
Direct output to
.IR path ,
rather than playing audio on the native audio device. The format of the output
is specified by
.I type
which can be any of the supported output formats (see
.B Output Formats
below.) If a format is not specified, one will be inferred from
.IR path .
If
.I path
is a single dash (\-), the output will be written to standard output.
.TP
\fB\-b\fR or \fB\-\-bit\-depth=\fIdepth\fR
Request an output precision of
.I depth
bits per sample. Higher bit depths yield higher quality sound. Typical bit
depths are 8, 16, 24, and 32, however other depths may also be possible. See
the
.B NOTES
section below for further details about this option.
.TP
.BR \-d " or " \-\-no\-dither
Do not dither output PCM samples. This may result in lower sound quality but
is useful for analyzing output from the decoder.
.TP
\fB\-\-fade\-in\fR[\fB=\fIduration\fR]
Gradually fade-in the audio from each file over
.IR duration .
If not specified, the default duration is
.B 0:05
(five seconds.)
.TP
\fB\-a\fR or \fB\-\-attenuate=\fIdecibels\fR or \fB\-\-amplify=\fIdecibels\fR
Attenuate or amplify the signal by
.I decibels
(dB).
The signal is attenuated if the decibel value is negative; it is amplified if
the value is positive.
The decibel value must be in the range \-175 to +18.
The value may be fractional, e.g. \-1.5.
A value of 0 will leave the signal unchanged.
Each step of 6\ dB will approximately halve (in the negative direction) or
double (in the positive direction) the strength of the signal.
.SS Channel Selection
.PP
For dual channel streams, an output channel should be selected. If one is not
selected, the first (left) channel will be used.
.PP
For stereo streams, making a channel selection other than stereo will cause
the output to become monaural.
.TP
.BR \-1 " or " \-\-left
Output the first (left) channel only.
.TP
.BR \-2 " or " \-\-right
Output the second (right) channel only.
.TP
.BR \-m " or " \-\-mono
Mix the left and right channels together.
.TP
.BR \-S " or " \-\-stereo
Force stereo output, even if the stream is single or dual channel.
.SS Playback
.TP
\fB\-s\fR or \fB\-\-start=\fItime\fR
Begin playing at
.IR time ,
given as an offset from the beginning of the first file
.RB ( 0:00:00 ),
seeking as necessary.
.TP
\fB\-t\fR or \fB\-\-time=\fIduration\fR
Stop playback after the playing time of the output audio equals
.IR duration .
.TP
.BR \-z " or " \-\-shuffle
Randomize the list of files given on the command line for playback.
.TP
\fB\-r\fR or \fB\-\-repeat\fR[\fB=\fImax\fR]
Play the input files
.I max
times, or indefinitely. Playback can be stopped prematurely by giving a time
limit with the
.B \-t
.RB ( \-\-time )
option. If
.B \-z
.RB ( \-\-shuffle )
is also used, the files will be continuously shuffled and repeated in such a
way that the same file is not played again until at least half of the other
files have played in the interim.
.TP
.B \-\-tty\-control
Enable keyboard controls during playback. This is the default unless standard
input is not a terminal or output is redirected with
.B \-o
.RB ( \-\-output ).
The keyboard controls are:
.RS
.TP 3
.B P
Pause; press any key to resume.
.TP 3
.B S
Stop; press any key to replay the current file from the beginning.
.TP 3
.B F
Forward; advance to the next file.
.TP 3
.B B
Back; replay the current file, unless it has been playing for less than 4
seconds, in which case replay the previous file.
.TP 3
.B T
Time display; change the time display mode. This only works with
.B \-v
.RB ( \-\-verbose ).
The display mode alternates among overall playing time, current time
remaining, and current playing time.
.TP 3
.B Q
Quit; stop decoding and exit.
.RE
.TP
.B \-\-no\-tty\-control
Disable keyboard controls during playback. This is the default when standard
input is not a terminal or output is redirected with
.B \-o
.RB ( \-\-output ).
.SS Miscellaneous
.TP
.BR \-V " or " \-\-version
Display the effective version and build options for
.B madplay
and exit.
.TP
.B \-\-license
Display a license message and exit.
.TP
.BR \-h " or " \-\-help
Display a usage message and exit.
.SH Output Formats
Other than playing on the native audio device, the following output formats
are supported:
.TP
.B cdda
CD audio, 16-bit 44100 Hz stereo PCM
.RB ( *.cdr ,
.BR *.cda )
.TP
.B aiff
Audio IFF, [16-bit] PCM
.RB ( *.aif ,
.BR *.aiff )
.TP
.B wave
Microsoft RIFF/WAVE, [16-bit] PCM
.RB ( *.wav )
.TP
.B snd
Sun/NeXT audio, 8-bit ISDN \(*m-law
.RB ( *.au ,
.BR *.snd )
.TP
.B raw
binary [16-bit] host-endian linear PCM, stereo interleaved
.TP
.B hex
ASCII hexadecimal [24-bit] linear PCM, stereo interleaved, one sample per
output line (debug builds only)
.TP
.B esd
Enlightened Sound Daemon (EsounD) [16-bit] (give speaker host as
.IR path )
.TP
.B null
no output (usually for testing or timing the decoder)
.PP
Default bit depths shown in square brackets can be changed with the
.B \-b
.RB ( \-\-bit\-depth )
option.
.PP
Note that EsounD support requires the
.I libesd
library.
.SH Time Specifications
For options which accept a time or duration argument, the following time
specifications are recognized:
.TP
.IB hh : mm : ss . ddd
Hours, minutes, seconds, and decimal fractions of a second. This specification
is flexible;
.IB hh : mm : ss\c
,
.IB mmm : ss\c
,
.BI : ss\c
,
.IB sss . ddd\c
,
.BI . ddd\c
, and
.I ssss
are all acceptable. The component values are not constrained to any particular
range or number of digits.
.TP
.IB frac / unit
A length of time specified as a rational number, in seconds. This can be used
for sample-granularity, for example
.B 32/44100
for 32 samples, assuming a 44100 Hz sample frequency.
.TP
.IB time1 + time2
A composite time made by adding two time values together. This permits mixing
the above specification forms.
.PP
The resolution of any time value cannot exceed 1/352800000 seconds.
.SH DIAGNOSTICS
.TP
.B error: frame #: lost synchronization
If encountered at the beginning of a file, this means the file contains
something other than an ID3v2 tag before the MPEG audio data. If encountered
in the middle of a file, it may mean the file is corrupt. This message is most
commonly encountered, however, at the end of a file if the file contains an
ID3v1 tag that is not aligned to an MPEG audio frame boundary. In this case,
the message is harmless and may be ignored.
.PP
.TP
.B error: frame #: bad main_data_begin pointer
This message can occur while decoding a Layer\ III stream that has been cut or
spliced without preserving its bit reservoir. The affected frame cannot be
properly decoded, but it will be used to help restore the bit reservoir for
the next frame.
.PP
Most other messages indicate a deficiency in the input stream.
.PP
When a frame cannot be properly decoded, a concealment strategy is used as
follows:
.TP 2
\(bu
If the previous frame was properly decoded, it is repeated in place of the
current frame.
.TP 2
\(bu
If the previous frame was
.I not
properly decoded, the current frame is muted.
.SH CONFORMING TO
MAD conforms to Part\ 3 of the ISO/IEC\ 11172 (MPEG-1) international standard
for decoding MPEG audio. In addition, MAD supports the extension to Lower
Sampling Frequencies (LSF) as defined in Part\ 3 of ISO/IEC\ 13818 (MPEG-2).
.PP
The output from MAD has been tested and found to satisfy the ISO/IEC\ 11172-4
computational accuracy requirements for compliance. In most configurations,
MAD is a
.I Full Layer\ III ISO/IEC\ 11172-3 audio decoder
as defined by the standard.
.PP
The ID3 tag parsing library used by
.B madplay
conforms to the ID3v2.4.0 informal standard.
.SH NOTES
Because MAD produces samples with a precision greater than 24 bits, by default
.B madplay
will dither the samples to the precision of the output format. This produces
high quality audio that generally sounds superior to the output of a simple
rounding algorithm. However, dithering may unfavorably affect an analytic
examination of the output, and therefore it may be disabled by using the
.B \-d
.RB ( \-\-no\-dither )
option.
.PP
The actual precision of output samples can be requested with the
.B \-b
.RB ( \-\-bit\-depth )
option. Whether this request can be honored depends on the output format. If
output is directed to an audio device, success depends on the capabilities of
the device. If this option is not specified, a typical default depth will be
used (often 16) or in the case of output to an audio device, the highest bit
depth determined to work reliably with the device will be used.
.PP
Note that bit depths greater than 24 are effectively the same as 24-bit
precision samples padded to the requested depth.
.SH BUGS
The granularity of start and stop times
.RB ( \-\-start
and
.BR \-\-time )
is not yet as fine as this document suggests.
.SH AUTHOR
Robert Leslie <rob@mars.org>
.SH SEE ALSO
.BR sox (1),
.BR wget (1)
.\" .BR libmad (3)


consumer/mad/mad-0.14.2b/madplay.c
/*
 * mad - MPEG audio decoder
 * Copyright (C) 2000-2001 Robert Leslie
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id: madplay.c,v 1.67 2001/11/09 00:58:45 rob Exp $
 */

# ifdef HAVE_CONFIG_H
#  include "config.h"
# endif

# include "global.h"

/* include this first to avoid conflicts with MinGW __argc et al. */
# include "getopt.h"

# include <locale.h>
# include <stdio.h>
# include <stdarg.h>
# include <stdlib.h>
# include <string.h>

# ifdef HAVE_ASSERT_H
#  include <assert.h>
# endif

# ifdef HAVE_UNISTD_H
#  include <unistd.h>
# endif

# include <ctype.h>
# include <math.h>

# include "version.h"
# include "mad.h"
# include "audio.h"
# include "player.h"

# define FADE_DEFAULT	"0:05"

# if defined(EXPERIMENTAL)
static int external_mix;
static int experimental;
# endif

static
struct option const options[] = {
  { "amplify",		required_argument, 0,		 'a' },
  { "attenuate",	required_argument, 0,		 'a' },
  { "author",		no_argument,	   0,		-'a' },
  { "bit-depth",	required_argument, 0,		 'b' },
  { "downsample",	no_argument,	   0,		-'d' },
  { "fade-in",		optional_argument, 0,		-'i' },
  { "help",		no_argument,	   0,		 'h' },
  { "ignore-crc",	no_argument,	   0,		 'i' },
  { "left",		no_argument,	   0,		 '1' },
  { "license",		no_argument,	   0,		-'l' },
  { "mono",		no_argument,	   0,		 'm' },
  { "no-dither",	no_argument,	   0,		 'd' },
  { "output",		required_argument, 0,		 'o' },
  { "quiet",		no_argument,	   0,		 'q' },
  { "repeat",		optional_argument, 0,		 'r' },
  { "right",		no_argument,	   0,		 '2' },
  { "shuffle",		no_argument,	   0,		 'z' },
  { "start",		required_argument, 0,		 's' },
  { "stereo",		no_argument,	   0,		 'S' },
  { "time",		required_argument, 0,		 't' },
  { "verbose",		no_argument,	   0,		 'v' },
  { "version",		no_argument,	   0,		 'V' },
  { "very-quiet",	no_argument,	   0,		 'Q' },
# if defined(USE_TTY)
  { "tty-control",	no_argument,	   0,		-'c' },
  { "no-tty-control",	no_argument,	   0,		-'C' },
# endif
# if 0
  { "cross-fade",	no_argument,	   0,		 'x' },
  { "fade-out",		optional_argument, 0,		-'o' },
  { "gap",		required_argument, 0,		 'g' },
# endif
# if defined(EXPERIMENTAL)
  { "external-mix",	no_argument,	   &external_mix,  1 },
  { "experimental",	no_argument,	   &experimental,  1 },
# endif
  { 0 }
};

char const *argv0;

# define EPUTS(str)	fputs((str), stderr)

/*
 * NAME:	show_usage()
 * DESCRIPTION:	display usage message
 */
static
void show_usage(int verbose)
{
  fprintf(stderr, _("Usage: %s [OPTIONS] FILE [...]\n"), argv0);

  if (!verbose) {
    fprintf(stderr, _("Try `%s --help' for more information.\n"), argv0);
    return;
  }

  EPUTS(_("Decode and play MPEG audio FILE(s).\n"));

  /* the following usage text should agree with the option names */

  EPUTS(_("\nVerbosity:\n"));
  EPUTS(_("  -v, --verbose              show status while decoding\n"));
  EPUTS(_("  -q, --quiet                be quiet but show warnings\n"));
  EPUTS(_("  -Q, --very-quiet           be quiet and do not show warnings\n"));

  EPUTS(_("\nDecoding:\n"));
  EPUTS(_("      --downsample           reduce output sample rate 2:1\n"));
  EPUTS(_("  -i, --ignore-crc           ignore CRC errors\n"));

  EPUTS(_("\nAudio output:\n"));
  EPUTS(_("  -o, --output=[TYPE:]PATH   send output to PATH with format TYPE"
	                              " (see below)\n"));
  EPUTS(_("  -b, --bit-depth=DEPTH      request DEPTH bits per sample\n"));
  EPUTS(_("  -d, --no-dither            do not dither output PCM samples\n"));
  fprintf(stderr,
	_("      --fade-in[=DURATION]   fade-in songs over DURATION"
	                              " (default %s)\n"), FADE_DEFAULT);
# if 0
  fprintf(stderr,
	_("      --fade-out[=DURATION]  fade-out songs over DURATION"
	                              " (default %s)\n"), FADE_DEFAULT);
  EPUTS(_("  -g, --gap=DURATION         set inter-song gap to DURATION\n"));
  EPUTS(_("  -x, --cross-fade           cross-fade songs"
	                              " (use with negative gap)\n"));
# endif
  EPUTS(_("  -a, --attenuate=DECIBELS   attenuate signal by DECIBELS (-)\n"));
  EPUTS(_("  -a, --amplify=DECIBELS     amplify signal by DECIBELS (+)\n"));

  EPUTS(_("\nChannel selection:\n"));
  EPUTS(_("  -1, --left                 output first (left) channel only\n"));
  EPUTS(_("  -2, --right                output second (right) channel"
	                              " only\n"));
  EPUTS(_("  -m, --mono                 mix left and right channels"
	                              " for monaural output\n"));
  EPUTS(_("  -S, --stereo               force stereo output\n"));

# if defined(EXPERIMENTAL)
  EPUTS(_("\nExperimental:\n"));
  EPUTS(_("      --external-mix         output pre-synthesis samples"
	                              " for external mixer\n"));
  EPUTS(_("      --experimental         enable experimental filter\n"));
# endif

  EPUTS(_("\nPlayback:\n"));
  EPUTS(_("  -s, --start=TIME           skip to begin at TIME"
	                              " (HH:MM:SS.DDD)\n"));
  EPUTS(_("  -t, --time=DURATION        play only for DURATION"
	                              " (HH:MM:SS.DDD)\n"));
  EPUTS(_("  -z, --shuffle              randomize file list\n"));
  EPUTS(_("  -r, --repeat[=MAX]         play files MAX times,"
	                              " or indefinitely\n"));
# if defined(USE_TTY)
  EPUTS(_("      --tty-control          enable keyboard controls\n"));
  EPUTS(_("      --no-tty-control       disable keyboard controls\n"));
# endif

  EPUTS(_("\nMiscellaneous:\n"));
  EPUTS(_("  -V, --version              display version number and exit\n"));
  EPUTS(_("      --license              show copyright/license message"
	                              " and exit\n"));
  EPUTS(_("  -h, --help                 display this help and exit\n"));

  EPUTS(_("\nSupported output formats:\n"));
  EPUTS(_("  cdda    CD audio, 16-bit 44100 Hz stereo PCM (*.cdr, *.cda)\n"));
  EPUTS(_("  aiff    Audio IFF, [16-bit] PCM (*.aif, *.aiff)\n"));
  EPUTS(_("  wave    Microsoft RIFF/WAVE, [16-bit] PCM (*.wav)\n"));
  EPUTS(_("  snd     Sun/NeXT audio, 8-bit ISDN mu-law (*.au, *.snd)\n"));
  EPUTS(_("  raw     binary [16-bit] host-endian linear PCM\n"));
# if defined(DEBUG)
  EPUTS(_("  hex     ASCII hexadecimal [24-bit] linear PCM\n"));
# endif
# if defined(HAVE_LIBESD)
  EPUTS(_("  esd     Enlightened Sound Daemon [16-bit]"
	  " (give speaker host as PATH)\n"));
# endif
  EPUTS(_("  null    no output (decode only)\n"));
}

# undef EPUTS

/*
 * NAME:	verror()
 * DESCRIPTION:	print error message with program title prefix
 */
static
void verror(char const *message, va_list args)
{
  fprintf(stderr, "%s: ", argv0);
  vfprintf(stderr, message, args);
  fputc('\n', stderr);
}

/*
 * NAME:	warn()
 * DESCRIPTION:	print warning message
 */
static
void warn(char const *message, ...)
{
  va_list args;

  va_start(args, message);
  verror(message, args);
  va_end(args);
}

/*
 * NAME:	die()
 * DESCRIPTION:	exit with failure status after printing message
 */
static
void die(char const *message, ...)
{
  va_list args;

  va_start(args, message);
  verror(message, args);
  va_end(args);

  exit(1);
}

/*
 * NAME:	parse_time()
 * DESCRIPTION:	parse a time specification string
 */
static
int parse_time(mad_timer_t *timer, char const *str)
{
  mad_timer_t time, accum = mad_timer_zero;
  signed long decimal;
  unsigned long seconds, fraction, fracpart;
  int minus;

  while (isspace(*str))
    ++str;

  do {
    seconds = fraction = fracpart = 0;

    switch (*str) {
    case '-':
      ++str;
      minus = 1;
      break;

    case '+':
      ++str;
    default:
      minus = 0;
    }

    do {
      decimal = strtol(str, (char **) &str, 10);
      if (decimal < 0)
	return -1;

      seconds += decimal;

      if (*str == ':') {
	seconds *= 60;
	++str;
      }
    }
    while (*str >= '0' && *str <= '9');

    if (*str == '.'
# if defined(HAVE_LOCALECONV)
	|| *str == *localeconv()->decimal_point
# endif
	) {
      char const *ptr;

      decimal = strtol(++str, (char **) &ptr, 10);
      if (decimal < 0)
	return -1;

      fraction = decimal;

      for (fracpart = 1; str != ptr; ++str)
	fracpart *= 10;
    }
    else if (*str == '/') {
      decimal = strtol(++str, (char **) &str, 10);
      if (decimal < 0)
	return -1;

      fraction = seconds;
      fracpart = decimal;

      seconds  = 0;
    }

    mad_timer_set(&time, seconds, fraction, fracpart);
    if (minus)
      mad_timer_negate(&time);

    mad_timer_add(&accum, time);
  }
  while (*str == '-' || *str == '+');

  while (isspace(*str))
    ++str;

  if (*str != 0)
    return -1;

  *timer = accum;

  return 0;
}

/*
 * NAME:	get_time()
 * DESCRIPTION:	parse a time value or die
 */
static
mad_timer_t get_time(char const *str, int positive, char const *name)
{
  mad_timer_t time;

  if (parse_time(&time, str) == -1)
    die(_("invalid %s specification \"%s\""), name, str);

  if (positive && mad_timer_sign(time) <= 0)
    die(_("%s must be positive"), name);

  return time;
}

/*
 * NAME:	parse_decibels()
 * DESCRIPTION:	parse a decibel value specification string
 */
static
int parse_decibels(double *db, char const *str)
{
  *db = strtod(str, (char **) &str);

  while (isspace(*str))
    ++str;

  if (strncasecmp(str, "dB", 2) == 0) {
    str += 2;

    while (isspace(*str))
      ++str;
  }

  if (*str != 0)
    return -1;

  return 0;
}

/*
 * NAME:	get_decibels()
 * DESCRIPTION:	parse a decibel value into a linear ratio or die
 */
static
mad_fixed_t get_decibels(char const *str)
{
  double db;

  enum {
    DB_MIN = -175,
    DB_MAX =  +18
  };

  if (parse_decibels(&db, str) == -1)
    die(_("invalid decibel specification \"%s\""), str);

  if (db < DB_MIN || db > DB_MAX)
    die(_("decibel value must be in the range %+d to %+d"), DB_MIN, DB_MAX);

  return mad_f_tofixed(pow(10, db / 20));
}

/*
 * NAME:	get_options()
 * DESCRIPTION:	parse command-line options or die
 */
static
void get_options(int argc, char *argv[], struct player *player)
{
  int opt, index;
  int ttyset = 0;

  while ((opt = getopt_long(argc, argv,
			    "vqQ"	/* verbosity options */
			    "i"		/* decoding options */
			    "o:b:da:"	/* audio output options */
# if 0
			    "g:x"
# endif
			    "12mS"	/* channel selection options */
			    "s:t:zr::"	/* playback options */
			    "Vh",	/* miscellaneous options */
			    options, &index)) != -1) {
    switch (opt) {
    case 0:
      break;

    case '1':
    case '2':
      player->output.select = PLAYER_CHANNEL_LEFT + (opt - '1');
      break;

    case 'a':
      player->output.attenuation = get_decibels(optarg);
      break;

    case -'a':
      printf("%s\n", mad_author);
      exit(0);

    case 'b':
      opt = atoi(optarg);
      if (opt <= 0)
	die(_("invalid bit depth \"%s\""), optarg);

      player->output.precision_in = opt;
      break;

# if defined(USE_TTY)
    case -'c':
      player->options |= PLAYER_OPTION_TTYCONTROL;
      ttyset = 1;
      break;

    case -'C':
      player->options &= ~PLAYER_OPTION_TTYCONTROL;
      ttyset = 1;
      break;
# endif

    case 'd':
      player->output.mode = AUDIO_MODE_ROUND;
      break;

    case -'d':
      player->options |= PLAYER_OPTION_DOWNSAMPLE;
      break;

# if 0
    case 'g':
      player->gap = get_time(optarg, 0, _("gap time"));
      player->options |= PLAYER_OPTION_GAP;
      break;
# endif

    case 'h':
      show_usage(1);
      exit(0);

    case 'i':
      player->options |= PLAYER_OPTION_IGNORECRC;
      break;

    case -'i':
      player->fade_in = get_time(optarg ? optarg : FADE_DEFAULT, 1,
				 _("fade-in time"));
      player->options |= PLAYER_OPTION_FADEIN;
      break;

    case -'l':
      ver_license(stdout);
      exit(0);

    case 'm':
      player->output.select = PLAYER_CHANNEL_MONO;
      break;

    case 'o':
      if (player->output.path)
	die(_("multiple output destinations not supported"));

      player->output.path = optarg;

      player->output.command = audio_output(&player->output.path);
      if (player->output.command == 0)
	die(_("unknown output format type for \"%s\""), optarg);

      if (!ttyset)
	player->options &= ~PLAYER_OPTION_TTYCONTROL;
      break;

# if 0
    case -'o':
      player->fade_out = get_time(optarg ? optarg : FADE_DEFAULT, 1,
				  _("fade-out time"));
      player->options |= PLAYER_OPTION_FADEOUT;
      break;
# endif

    case 'q':
      player->verbosity = -1;
      break;

    case 'Q':
      player->verbosity = -2;
      break;

    case 'r':
      if (optarg == 0)
	player->repeat = -1;
      else {
	player->repeat = atoi(optarg);
	if (player->repeat <= 0)
	  die(_("invalid repeat count \"%s\""), optarg);
      }
      break;

    case 's':
      player->global_start = get_time(optarg, 0, _("start time"));
      player->options |= PLAYER_OPTION_SKIP;
      break;

    case 'S':
      player->output.select = PLAYER_CHANNEL_STEREO;
      break;

    case 't':
      player->global_stop = get_time(optarg, 1, _("playing time"));
      player->options |= PLAYER_OPTION_TIMED;
      break;

    case 'v':
      player->verbosity = +1;
      break;

    case 'V':
      ver_banner(stdout);
      fprintf(stderr, _("`%s --license' for licensing information.\n"), argv0);
      fprintf(stderr, _("Build options: %s\n"), mad_build);
      exit(0);

# if 0
    case 'x':
      player->options |= PLAYER_OPTION_CROSSFADE;
      break;
# endif

    case 'z':
      player->options |= PLAYER_OPTION_SHUFFLE;
      break;

    case '?':
      show_usage(0);
      exit(1);

    default:
      assert(!"option handler");
    }
  }

  if (optind == argc) {
    show_usage(0);
    exit(2);
  }
}

/*
 * NAME:	main()
 * DESCRIPTION:	program entry point
 */
int main(int argc, char *argv[])
{
  struct player player;
  int result = 0;

  argv0 = argv[0];

  /* internationalization support */

# if defined(ENABLE_NLS)
  setlocale(LC_ALL, "");
  bindtextdomain(PACKAGE, LOCALEDIR);
  textdomain(PACKAGE);
# endif

  /* initialize and get options */

  player_init(&player);

# if !defined(__CYGWIN__)  /* Cygwin support for this is currently buggy */
  /* check for default tty control */
  if (isatty(STDIN_FILENO))
    player.options |= PLAYER_OPTION_TTYCONTROL;
# endif

  get_options(argc, argv, &player);

  /* main processing */

  if (player.verbosity >= 0)
    ver_banner(stderr);

  if (player.options & PLAYER_OPTION_CROSSFADE) {
    if (!(player.options & PLAYER_OPTION_GAP))
      warn(_("cross-fade ignored without gap"));
    else if (mad_timer_sign(player.gap) >= 0)
      warn(_("cross-fade ignored without negative gap"));
  }

  /* make stop time absolute */
  if (player.options & PLAYER_OPTION_TIMED)
    mad_timer_add(&player.global_stop, player.global_start);

  /* get default audio output module */
  if (player.output.command == 0)
    player.output.command = audio_output(0);

# if defined(EXPERIMENTAL)
  if (external_mix) {
    player.options |= PLAYER_OPTION_EXTERNALMIX;
    player.output.command = 0;
  }
  if (experimental)
    player.options |= PLAYER_OPTION_EXPERIMENTAL;
# endif

  /* run the player */

  if (player_run(&player, argc - optind, (char const **) &argv[optind]) == -1)
    result = 4;

  /* finish up */

  player_finish(&player);

  return result;
}
